### PR TITLE
Add Do Not Disturb Applet

### LINKED ISSUE #166: Add a menu item to toggle "do-not-disturb"
Feature request : add a menu item to toggle "do-not-disturb" instead of having to do this through the settings window.

#### Expected behaviour

Very often when I'm at work I want to easily & quickly toggle the "do-not-disturb" feature.

#### Actual behaviour

To do so today, I have to click on the "Settings..." item of the applet to open up the "Indicator Notificatino Settings" windows and then check the "Enable do not disturb" checkbox.

#### MATE general version

1.22

#### Package version

1.22.1-0ubuntu1

#### Linux Distribution

Ubuntu-Mate 19.10

#### Link to downstream report of your Distribution

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/89257609-add-a-menu-item-to-toggle-do-not-disturb?utm_campaign=plugin&utm_content=tracker%2F1172660&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1172660&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by rbuj**
closes #166

### CODE REVIEW

**Review by lukefromdc — APPROVED** [review]
Build finishes and applet adds to the panel fine. Had to poke around a bit to figure out how to use it (you turn notifications on and off from the menu), but this does work. Icon could be a bit more descriptive but that could be a later commit. Anyone installing this in the panel themselves will know what it is, but if distros later add it by default it will need to be more discoverable.

About the UI: There are two ways this could be done: select do-not-disturb state with a menu (needed if the icon is not discriptive to prevent unintended state changes) or from a button directly toggling the do-not-disturb state, with menu on right-click.

We have lots of applets and things working both way, as I recall that was one of Ubuntu's complaints about the tray circa 2010 that led them to work on indicators (everything using a menu in the tray), but for some things like volume controls that made the UI less clear. Indicator volume slider as a menu item is horizontal for instance. Thus we see cases for both button and for menu UI's. Approving as it does work and is a good start on a needed feature.

**Comment by raveit65** [comment]
```
configure: error: Package requirements (mate-desktop-2.0 >= 1.17.0 libmatepanelapplet-4.0 >= 1.17.0 gtk+-3.0 >= 3.22.0 glib-2.0 >= 2.50.0) were not met:
```
https://travis-ci.org/github/mate-desktop/mate-notification-daemon/jobs/744153676#L897

You should add `mate-pannel-devel` build requires to`.build.yml`. See mate-applets for package names.

**Comment by rbuj** [comment]
done

**Comment by raveit65** [comment]
Distcheck failed.
```
make[4]: *** No rule to make target 'mate-notification-applet-menu.xml', needed by 'mate-notification-applet-resources.h'.  Stop.
```
https://travis-ci.org/github/mate-desktop/mate-notification-daemon/jobs/744494506#L1905

**Comment by rbuj** [comment]
done

File: src/capplet/mate-notification-applet.c [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+  { "About", "help-about", N_("_About"),
+    NULL, NULL, G_CALLBACK (show_about) }
+};
+
+static const char* notification_properties[] = {
+  "mate-notification-properties",
+};
+
+static void
+applet_destroy (MatePanelApplet        *applet_widget,
```
Comment (raveit65) [inline src/capplet/mate-notification-applet.c:62]: Build warning.
https://travis-ci.org/github/mate-desktop/mate-notification-daemon/jobs/744844272#L2245
```
mate-notification-applet.c:62:41: warning: unused parameter 'applet_widget' [-Wunused-parameter]

applet_destroy (MatePanelApplet        *applet_widget,
```

File: src/capplet/mate-notification-applet.c [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+{
+  g_assert (applet);
+
+  g_object_unref (applet->settings);
+  g_object_unref (applet->action_group);
+  g_free (applet);
+}
+
+static void
+call_properties (GtkAction              *action,
```
Comment (raveit65) [inline src/capplet/mate-notification-applet.c:73]: https://travis-ci.org/github/mate-desktop/mate-notification-daemon/jobs/744844272#L2248
```
mate-notification-applet.c:73:42: warning: unused parameter 'action' [-Wunused-parameter]

call_properties (GtkAction              *action,
```

File: src/capplet/mate-notification-applet.c [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+      mate_gdk_spawn_command_line_on_screen (gtk_widget_get_screen (GTK_WIDGET (applet->applet)),
+                                             notification_properties[i],
+                                             NULL);
+      return;
+    }
+  }
+}
+
+static void
+show_about (GtkAction              *action,
```
Comment (raveit65) [inline src/capplet/mate-notification-applet.c:92]: https://travis-ci.org/github/mate-desktop/mate-notification-daemon/jobs/744844272#L2251
```
mate-notification-applet.c:92:37: warning: unused parameter 'action' [-Wunused-parameter]

show_about (GtkAction              *action,
```

File: src/capplet/mate-notification-applet.c [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            MateNotificationApplet *applet)
```
Comment (raveit65) [inline src/capplet/mate-notification-applet.c:93]: https://travis-ci.org/github/mate-desktop/mate-notification-daemon/jobs/744844272#L2254
```
mate-notification-applet.c:93:37: warning: unused parameter 'applet' [-Wunused-parameter]

            MateNotificationApplet *applet)
```

I am still wondering why both parameters are unused 0_o

File: src/capplet/mate-notification-applet.c [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+  g_signal_connect (G_OBJECT (applet->settings), "changed::" GSETTINGS_KEY_DO_NOT_DISTURB,
+                    G_CALLBACK (settings_changed), applet);
+
+  return TRUE;
+}
+
+static gboolean
+applet_factory (MatePanelApplet *applet,
+                const gchar     *iid,
+                gpointer         data)
```
Comment (raveit65) [inline src/capplet/mate-notification-applet.c:201]: https://travis-ci.org/github/mate-desktop/mate-notification-daemon/jobs/744844272#L2353
```
mate-notification-applet.c:201:34: warning: unused parameter 'data' [-Wunused-parameter]

                gpointer         data)
```

File: src/capplet/Makefile.am [hunk 6]
```diff
@@ -9,6 +9,7 @@ NULL =
 
 mate_notification_properties_CFLAGS = \
 	-I$(top_srcdir) \
+	-I$(top_srcdir)/src/common \
 	-I$(top_srcdir)/src/daemon \
```
Comment (raveit65) [inline src/capplet/Makefile.am:13]: Why this line is twice?

File: src/capplet/mate-notification-applet.c (hunk 4 above)
Comment (rbuj) [reply src/capplet/mate-notification-applet.c:93]: It's a callback method.

**Comment by rbuj** [comment]
@raveit65 done: unused argument warning.

File: src/capplet/mate-notification-applet.c [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+  mate_panel_applet_setup_menu_from_resource  (applet->applet,
+                                               RESOURCE_PATH "menu.xml",
+                                               applet->action_group);
+
+  g_settings_bind (applet->settings, "do-not-disturb",
+                   do_not_disturb_toggle_action, "active",
+                   G_SETTINGS_BIND_DEFAULT);
+
+  g_signal_connect(G_OBJECT (applet->applet), "destroy",
```
Comment (raveit65) [inline src/capplet/mate-notification-applet.c:189]: A space is missing before the bracket.
`g_signal_connect (G_OBJECT (applet->applet), "destroy",`

File: src/capplet/mate-notification-applet.c [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+  g_object_unref (applet->action_group);
+  g_free (applet);
+}
+
+static void
+call_properties (GtkAction              *action,
+                 MateNotificationApplet *applet)
+{
+  gsize i;
+  (void)action;
```
Comment (raveit65) [inline src/capplet/mate-notification-applet.c:76]: A space is missing, or not?
`(void) action;`

File: src/capplet/mate-notification-applet.c [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+show_about (GtkAction              *action,
+            MateNotificationApplet *applet)
+{
+  static const char *authors[] = {
+    "Robert Buj <robert.buj@gmail.com>",
+    NULL
+  };
+  (void)action;
+  (void)applet;
+
```
Comment (raveit65) [inline src/capplet/mate-notification-applet.c:101]: Same here?

File: src/capplet/mate-notification-applet.c [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+  return TRUE;
+}
+
+static gboolean
+applet_factory (MatePanelApplet *applet,
+                const gchar     *iid,
+                gpointer         data)
+{
+  gboolean retval = FALSE;
+  (void)data;
```
Comment (raveit65) [inline src/capplet/mate-notification-applet.c:205]: And here?

**Comment by rbuj** [comment]
@raveit65 done - format code

**Review by raveit65 — APPROVED** [review]
LGTM,
new applet works fine.

**Comment by kbsali** [comment]
thanks @rbuj  & @raveit65

**Pull request merged**